src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/QueueClientDiagnosticsTests.cs

```diff
@@ -6,558 +6,558 @@ namespace Microsoft.Azure.ServiceBus.UnitTests.Diagnostics
-
-                int count = 0;
-                this.queueClient.RegisterMessageHandler((msg, ct) =>
+                try
+                {
+                    using (var listener = this.CreateEventListener(queueName, eventQueue))
+                    using (var subscription = this.SubscribeToEvents(listener))
                     {
-                        if (count++ == 0)
+                        await TestUtility.SendMessagesAsync(queueClient.InnerSender, 1);
+                        listener.Enable((name, queue, arg) => !name.EndsWith(".Start") && !name.Contains("Receive") );
+                                
+                        var count = 0;
+                        var exceptionCalled = false;
+                        var tcs = new TaskCompletionSource<int>(TaskCreationOptions.RunContinuationsAsynchronously);
+                
+                        queueClient.RegisterMessageHandler((msg, ct) =>
                         {
-                            throw new Exception("123");
-                        }
+                            if (count++ == 0)
+                            {
+                                throw new Exception("123");
+                            }
+                            tcs.TrySetResult(count);
+                            return Task.CompletedTask;
+                        },
+                        exArgs =>
+                        {
+                            // Do not set the completion source exception to avoid throwing 
+                            // when the task is awaited.  The sentinal variable is checked to detect
+                            // exception cases.
+                            exceptionCalled = true;
+                            return Task.CompletedTask;
+                        });
+                
+                        await tcs.Task.WithTimeout(TimeSpan.FromSeconds(MaxWaitSec));
+                        Assert.True(exceptionCalled);
 
-                        processingDone.Set();
-                        return Task.CompletedTask;
-                    },
-                    exArgs =>
-                    {
-                        exceptionCalled = true;
-                        return Task.CompletedTask;
-                    });
-                processingDone.WaitOne(TimeSpan.FromSeconds(maxWaitSec));
-                Assert.True(exceptionCalled);
-
-                // message is processed, but abandon happens after that
-                // let's spin until Complete call starts and ends
-                int wait = 0;
-                while (wait++ < maxWaitSec && this.events.Count < 3)
-                {
-                    await Task.Delay(TimeSpan.FromSeconds(1));
-                }
+                        // message is processed, but abandon happens after that
+                        // let's spin until Complete call starts and ends
+                        int wait = 0;
+                        while (wait++ < MaxWaitSec && eventQueue.Count < 3)
```

Comment:
What about extracting this into a helper?

Reply:
Good thought, and I'll take note for the next pass.  This one turned out to be a more substantial set of changes than I anticipated and I'd like to minimize the moving parts.